transport-classes-epoll/src/main/java/io/netty5/channel/epoll/NativeDatagramPacketArray.java

```diff
@@ -109,24 +107,39 @@ boolean addReadable(Buffer buf, int segmentLen, InetSocketAddress recipient) {
                 return false;
             }
             for (var c = first; c != null; c = c.nextReadable()) {
-                int writableBytes = c.readableBytes();
-                int byteCount = segmentLen == 0? writableBytes : Math.min(writableBytes, segmentLen);
-                if (iovArray.addReadable(c, byteCount)) {
-                    NativeDatagramPacket p = packets[count];
-                    long packetAddr = iovArray.memoryAddress(iovArrayStart);
-                    p.init(packetAddr, iovArray.count() - iovArrayStart, segmentLen, recipient);
-                    count++;
-                    c.skipReadableBytes(byteCount);
+                if (segmentLen == 0) {
+                    int readableBytes = c.readableBytes();
+                    if (iovArray.addReadable(c, readableBytes)) {
+                        c.skipReadableBytes(readableBytes);
+                    } else {
+                        break;
+                    }
+                } else {
+                    // In the case of segments we need to add each of the segments in a loop.
+                    while (c.readableBytes() > 0) {
+                        int byteCount = Math.min(c.readableBytes(), segmentLen);
+                        if (iovArray.addReadable(c, byteCount)) {
+                            c.skipReadableBytes(byteCount);
+                        } else {
+                            break;
+                        }
+                    }
                 }
+
             }
+            // Add one packet for sendmmsg(...) now.
+            NativeDatagramPacket p = packets[count];
+            long packetAddr = iovArray.memoryAddress(iovArrayStart);
+            p.init(packetAddr, iovArray.count() - iovArrayStart, segmentLen, recipient);
+            count++;
```

Comment:
I fixed two bugs here... so even if we don't merge this PR we want this change. I just noticed it with this PR as I did add some extra validation in the `WriteSink.complete(...)` methods which ensured we never reported more messages as written then what is in the outbound buffer.

Reply:
@chrisvest